libraries/chain/include/graphene/chain/betting_market_object.hpp

```diff
@@ -0,0 +1,716 @@
+      void on_closed_event(database& db);
+      void on_graded_event(database& db, betting_market_resolution_type new_grading);
+      void on_settled_event(database& db);
+      void on_canceled_event(database& db);
+   private:
+      class impl;
+      std::unique_ptr<impl> my;
+};
+
+class bet_object : public graphene::db::abstract_object< bet_object >
+{
+   public:
+      static const uint8_t space_id = protocol_ids;
+      static const uint8_t type_id = bet_object_type;
+
+      account_id_type bettor_id;
+      
+      betting_market_id_type betting_market_id;
+
+      asset amount_to_bet;
+
+      bet_multiplier_type backer_multiplier;
+
+      bet_type back_or_lay;
+
+      fc::optional<fc::time_point_sec> end_of_delay;
+
+      static share_type get_approximate_matching_amount(share_type bet_amount, bet_multiplier_type backer_multiplier, bet_type back_or_lay, bool round_up = false);
+
+      // returns the amount of a bet that completely matches this bet
+      share_type get_approximate_matching_amount(bool round_up = false) const;
+
+      static share_type get_exact_matching_amount(share_type bet_amount, bet_multiplier_type backer_multiplier, bet_type back_or_lay);
+      share_type get_exact_matching_amount() const;
+
+      static std::pair<share_type, share_type> get_ratio(bet_multiplier_type backer_multiplier);
+      std::pair<share_type, share_type> get_ratio() const; 
+
+      // returns the minimum amount this bet could have that could be matched at these odds
+      share_type get_minimum_matchable_amount() const; 
+      // returns the minimum amount another user could bet to match this bet at these odds
+      share_type get_minimum_matching_amount() const;
+};
+
+class betting_market_position_object : public graphene::db::abstract_object< betting_market_position_object >
+{
+   public:
+      static const uint8_t space_id = implementation_ids;
+      static const uint8_t type_id = impl_betting_market_position_object_type;
+
+      account_id_type bettor_id;
+      
+      betting_market_id_type betting_market_id;
+
+      share_type pay_if_payout_condition;
+      share_type pay_if_not_payout_condition;
+      share_type pay_if_canceled;
+      share_type pay_if_not_canceled;
+      share_type fees_collected;
+      
+      share_type reduce();
+};
+
+typedef multi_index_container<
+   betting_market_rules_object,
+   indexed_by<
+      ordered_unique< tag<by_id>, member< object, object_id_type, &object::id > >
+   > > betting_market_rules_object_multi_index_type;
+typedef generic_index<betting_market_rules_object, betting_market_rules_object_multi_index_type> betting_market_rules_object_index;
+
+typedef multi_index_container<
+   betting_market_group_object,
+   indexed_by<
+      ordered_unique< tag<by_id>, member< object, object_id_type, &object::id > >,
+      ordered_non_unique< tag<by_event_id>, member<betting_market_group_object, event_id_type, &betting_market_group_object::event_id> >,
```

Comment:
This should be made `ordered_unique` to enforce a well-defined order when iterating.

Reply:
I assume for these ones, we'd need to change the index slightly or otherwise would break existing behavior. Simply replacing `ordered_non_unique` by `ordered_unique` may not do it for us.